core/src/main/java/org/testcontainers/images/RemoteDockerImage.java

```diff
@@ -82,19 +83,20 @@ protected final String resolve() {
                 }
 
                 if (attempts++ >= 3) {
-                    logger.error("Retry limit reached while trying to pull image: " + dockerImageName + ". Please check output of `docker pull " + dockerImageName + "`");
-                    throw new ContainerFetchException("Retry limit reached while trying to pull image: " + dockerImageName);
+                    logger.error("Retry limit reached while trying to pull image: {}" + dockerImageName + ". Please check output of `docker pull {}`", dockerImageName, dockerImageName);
+                    throw new ContainerFetchException("Retry limit reached while trying to pull image: " + dockerImageName, lastException);
                 }
 
                 // The image is not available locally - pull it
                 try {
-                    dockerClient.pullImageCmd(dockerImageName).exec(new PullImageResultCallback()).awaitCompletion();
-                } catch (InterruptedException e) {
-                    throw new ContainerFetchException("Failed to fetch container image for " + dockerImageName, e);
+                    final PullImageResultCallback callback = new PullImageResultCallback();
+                    dockerClient.pullImageCmd(dockerImageName).exec(callback);
+                    callback.awaitSuccess();
+                    AVAILABLE_IMAGE_NAME_CACHE.add(dockerImageName);
+                    break;
+                } catch (DockerClientException e) {
+                    lastException = e;
                 }
-
-                // Do not break here, but step into the next iteration, where it will be verified with listImagesCmd().
-                // see https://github.com/docker/docker/issues/10708
```

Comment:
This issue is resolved as of Jan 2016 so I don't think we need to worry about it any more.

Accordingly, I've tidied the code a bit. Crucially the control flow now breaks out early if the image has been successfully pulled (line 96), instead of re-fetching the image listing.

This is actually important in letting us pull by sha256 hash: the listing doesn't include the same hash as the repository, so we can't use the listing to check for its presence (possibly something we can fix upstream in docker-java for better performance)

Reply:
Can we also add a test for pulling with sha256 hash?

Reply:
Done. While I was at it I tidied the image name handling code and added more tests relating to image names/image pulling.